source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.h

```diff
@@ -0,0 +1,208 @@
+ * sampleRTT calculation window. In this, all of the latency samples are consolidated into a
+ * configurable quantile value to represent the measured latencies. This quantile value sets
+ * sampleRTT and the concurrency limit is updated as described in the algorithm section above.
+ *
+ * When not in a sampling window, the controller is simply servicing the adaptive concurrency filter
+ * via the public functions.
+ *
+ * Locking:
+ * ========
+ * There are 2 mutually exclusive calculation windows, so update window mutex is held to prevent the
+ * overlap of these windows. It is necessary for a worker thread to know specifically if the
+ * controller is inside of a minRTT recalculation window during the recording of a latency sample,
+ * so this extra bit of information is stored in the recalculating_min_rtt_ boolean. The
+ * recalculating_min_rtt_ flag can only be set if inside of a minRTT calculation window.
+ *
+ * The histogram that aggregates latency samples, latency_sample_hist_, must be protected by a
+ * separate lock (the latency sample mutex). The additional lock is necessary because not only is
+ * the histogram mutated during both calculation windows, but also during latency samples triggered
+ * during the adaptive concurrency filter's header encoding step.
+ */
+class GradientController : public ConcurrencyController {
+public:
+  GradientController(GradientControllerConfigSharedPtr config, Event::Dispatcher& dispatcher,
+                     Runtime::Loader& runtime, const std::string& stats_prefix,
+                     Stats::Scope& scope);
+
+  // ConcurrencyController.
+  RequestForwardingAction forwardingDecision() override;
+  void recordLatencySample(std::chrono::nanoseconds rq_latency) override;
+  uint32_t concurrencyLimit() const override { return concurrency_limit_.load(); }
+
+private:
+  static GradientControllerStats generateStats(Stats::Scope& scope,
+                                               const std::string& stats_prefix);
+  void updateMinRTT();
+  std::chrono::microseconds processLatencySamplesAndClear()
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(latency_sample_mtx_);
+  uint32_t calculateNewLimit() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+  void setMinRTTSamplingWindow() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+  bool inMinRTTSamplingWindow() const { return recalculating_min_rtt_.load(); }
+  void resetSampleWindow() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+
+  const GradientControllerConfigSharedPtr config_;
+  Event::Dispatcher& dispatcher_;
+  Stats::Scope& scope_;
+  GradientControllerStats stats_;
+
+  // Ensures that the minRTT calculation window and the sample window (where the new concurrency
+  // limit is determined) do not overlap.
+  absl::Mutex update_window_mtx_;
+
+  std::atomic<bool> recalculating_min_rtt_;
+
+  // Protects the latency sample histogram during mutations.
+  absl::Mutex latency_sample_mtx_ ABSL_ACQUIRED_AFTER(update_window_mtx_);
```

Comment:
I might be missing something, but for simplicity can't we just use update_window_mtx_ in the sample storage case? None of the other locks are contended very much and I'm not sure it is worth it right now to have a second lock? It makes it harder to understand and we can deal with any contention/perf issues later?

Reply:
`latency_sample_mtx_` is just guarding the histogram and nothing else. The compiler will actually complain if you try to mess with the histogram without holding this lock, so I (personally) don't think it adds any cognitive load when it comes to the actual implementation. It also strikes me as low-risk for future maintenance because the compiler will complain.

Reply:
My point is that you already have another mutex you can use to guard the histogram, and in fact are already holding it in I think 2 of the 3 places you access the histogram. I don't see a compelling reason to add a 2nd mutex right now. Let's just use a single mutex for everything the code does right now and we can revisit later if we see contention.